File: clean_home_inventory_closet_data.py
# simple python script to restructure my home inventory closet data
# starting point is csv with box, category, sub category, item, color
# goal is to remove the word after the last comma in the item column and place it into the sub category column

# this is my attempt at a python script that has the exact same function as the perl script of the same name in this repository

# find nth occurrence of character
# params:
#   line = the string you're searching
#   char = the character you're looking for
#   n    = the occurrence of the character that you're interested in
def find_nth(line, char, n):
    i = 0
    index = 0
    while i <= n:
        index = line.find(char, index + 1)
        i += 1
        if i == n or index == -1:
            return index

local_directory = '/Users/Katelynn/Documents/'                  # replace with local location of the repository
clothes_new = "box,category,sub_category,details,color\n";      # header row for new file

# open the local file
with open(local_directory + 'home_inventory/clothes_download.csv', 'r') as file1:
    next(file1)                                                 # skip first line
    
    # loop through the original file
    for line in file1:
        if (1 == 1):
#        if (line.find("bright pink, leopard print") != -1):
            print line
            line = line.lower()                                     # make all letters lowercase for consistency
            arr = line.split(',')                                   # split the line into parts
            box = arr[0].upper()                                    # first bit is the box the item is in, also the only uppercase column
            category = arr[1]                                       # second bit is the top level category
            sub_category = '';
            item = '';
            color = '';

            if category == '':                                      # handle undefined items by moving sub_category to category
                category = arr[2];
            else:
                # get item bit
                item = line[find_nth(line, "\"", 1)+1:find_nth(line, "\"", 2)]
                arr_item = item.split(',')                          # split the item column into parts
                sub_category = arr_item[len(arr_item) - 1].strip()  # get the last piece of the item column and save it as the sub category

                # clean up item bit
                item = '';
                for i in range(len(arr_item) - 1):
                    item += arr_item[i].strip() + ", "
                item = item[:len(item)-1]
                item = item[:-1]

                # get color bit
                last_double_quote = find_nth(line, "\"", 4)
                print "ldq:", last_double_quote
                if (last_double_quote == -1):
                    color = arr[len(arr) - 1]
                    color = color.strip().replace("\"",'')
                else:                                                       # for when item and color both had commas in them originally
                    for i in range(len(arr) - 1, -1, -1):
                        if (arr[i - 1].find("\"") != -1):
                            if (arr[i].strip().replace("\"",'') == sub_category):
                                print "sub"
                                last_double_quote = i
                            else:
                                print "not sub"
                                last_double_quote = i - 1
                            break
                        else:
                            last_double_quote = i
                        print "i =", i
                        
                    while last_double_quote < len(arr):
                        print "arr[last_double_quote]", arr[last_double_quote], "| last_double_quote =", last_double_quote
                        color += arr[last_double_quote].strip().replace("\"", '') + ", "
                        last_double_quote += 1
                    color = color[:-2]
            clothes_new += box + "," + category + "," + sub_category + ",\"" + item + "\",\"" + color + "\"\n"
print clothes_new;

# create new local file and write the cleaned up data to it
with open(local_directory + 'home_inventory/clothes_new_python.csv', 'w') as file2:
    file2.write(clothes_new)

File: clean_home_inventory_closet_data.pl
#!/usr/bin/perl

# simple perl script to restructure my home inventory closet data
# starting point is csv with box, category, sub category, item, color
# goal is to remove the word after the last comma in the item column and place it into the sub category column

use strict;
use warnings;
use Data::Dumper;

my $local_directory = '/Users/Katelynn/Documents/';                     # replace with local location of the repository

# open the local file
open my $fh1, $local_directory . 'home_inventory/clothes_download.csv'; 

# loop through the original file
my $clothes_new = "box,category,sub_category,details,color\n";          # header row for new file
while (my $line = <$fh1>) {
    next if $. == 1;                                                    # skip first line
#    next unless $line =~ /sheer/ && $line =~ /blouse/ && $line =~ /pale/;
    $line = lc $line;                                                   # make all letters lowercase for consistency
    my @arr = split/,/,$line;                                           # split the line into parts
    my $box = uc $arr[0];                                               # first bit is the box the item is in, also the only uppercase column
    my $category = $arr[1];                                             # second bit is the top level category
    my $sub_category = '';
    my $item;
    my $color = '';
    
    # get item bit
    ($item) = $line =~ /"(.*?)"/;
    
    if (defined $item) {
        my @arr_item = split/,/,$item;                                  # split the item column into parts
        $sub_category = substr @arr_item[(scalar @arr_item - 1)], 1;    # get the last piece of the item column and save it as the sub category

        # clean up item bit
        $item = '';
        for (my $i = 0; $i < (scalar @arr_item) - 1; $i++) {
            $item .= $arr_item[$i] . ',';
        }
        chop $item;
        
        # get color bit
        my $last_double_quote = 0;
        for (my $i = scalar @arr - 1; $i > 0; $i--) { 
            if ($arr[$i] =~ /\"/ && $i != scalar @arr - 1) {            # find last array index that has a double quote and isn't the last item in the array
                if ($arr[$i-1] =~ /\"/) {                               # for when item and color both had commas in them originally
                    if ($arr[$i] =~ /$sub_category/) {
                        $last_double_quote = $i;
                    } else {
                        $last_double_quote = $i - 1;
                    }
                } else {
                    $last_double_quote = $i;
                }
                last;
            }
        }
        for (my $i = $last_double_quote + 1; $i < scalar @arr; $i++) {  # use it as the starting place for the color part
            # clean up color bit
            $arr[$i] =~ s/^\s+|\"|\s+$//g;                              # remove leading spaces | remove double quotes | remove trailing spaces
            $color .= $arr[$i] . ", "; 
        }
        $color = substr $color, 0, length($color) - 2;
    } else {                                                            # handle undefined items
        $category = $arr[2];                                            # move sub_category to category
        $item = '';
    }
    $clothes_new .= "$box,$category,$sub_category,\"$item\",\"$color\"\n";
}
close $fh1;

# create new local file and write the cleaned up data to it
open my $fh2, '>', $local_directory . 'home_inventory/clothes_new_perl.csv';
print $fh2 $clothes_new;
close $fh2;